exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSetBinding object after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSetBinding); err != nil {
+			r.Log.Error(err, "failed to patch config")
+		}
+	}()
+
+	// Initializes the maps in ClusterResourceSetBinding object if they are uninitialized.
+	InitClusterResourceSetBinding(clusterResourceSetBinding, clusterResourceSet)
+
+	var clusterResourceSetInMap addonsv1.ResourcesSetBinding
+	var ok bool
+	if clusterResourceSetInMap, ok = clusterResourceSetBinding.Spec.Bindings[clusterResourceSet.Name]; !ok {
+		// This should never be the case as InitClusterResourceSetBinding is adding the ClusterResourceSet to the map during initialization.
+		return errors.New("failed to find ClusterResourceSet " + clusterResourceSet.Name + " in ClusterResourceBinding")
+	}
+
+	errList := []error{}
+
+	// Iterate all resources and apply them to the cluster and update the resource status in the ClusterResourceSetBinding object.
+	for _, resource := range clusterResourceSet.Spec.Resources {
+		// If resource is already applied successfully and clusterResourceSet mod is "ApplyOnce", continue. (No need to check hash changes here)
+		resourceInMap := addonsv1.ResourceBinding{}
+		if resourceInMap, ok = clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)]; ok {
+			if clusterResourceSet.Spec.Strategy == string(addonsv1.ClusterResourceSetStrategyApplyOnce) && resourceInMap.Applied {
+				continue
+			}
+		}
+
+		// This is set here to cover unsuccessful apply scenarios (if the end of the loop is not reached)
+		resourceInMap = *GenerateResourceStatus(false, "")
+		clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)] = resourceInMap
+
+		// Only allow using resources (Secrets/Configmaps) in the same namespace with the cluster.
+		typedName := types.NamespacedName{Name: resource.Name, Namespace: cluster.Namespace}
+		dataList := make([][]byte, 0)
+
+		// Retrieve data in the resource as an array because there can be many key-value pairs in resource and all will be applied to the cluster.
+		switch resource.Kind {
+		case string(addonsv1.ConfigMapClusterResourceSetResourceKind):
+			// Set status in ClusterResourceSetBinding in case of early continue due to a failure. Do this for only supported kinds.
+			clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)] = addonsv1.ResourceBinding{
+				Hash:            "",
+				Applied:         false,
+				LastAppliedTime: &metav1.Time{Time: time.Now().UTC()},
+			}
+
+			resourceConfigMap, err := GetConfigMapFromNamespacedName(context.Background(), r.Client, typedName)
+			if err != nil {
+				logger.Error(err, "Failed to find ConfigMap resource", "ConfigMap name", typedName)
+				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.MissingResourceReason, clusterv1.ConditionSeverityWarning, err.Error())
```

Comment:
We have CRSBinding for detailed info but wanted to report in CRS object to for example if a resource is missing. It may be flickering but do we only log if we don't use conditions?

Reply:
Created an issue for this: https://github.com/kubernetes-sigs/cluster-api/issues/3291